File: 03.CPP/0016.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <limits>

class Solution {
public:
    int threeSumClosest(std::vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = nums.size();

        int bestSum = 0;
        int bestDiff = std::numeric_limits<int>::max();

        for (int i = 0; i < n; i++) {
            int left = i + 1;
            int right = n - 1;

            while (left < right) {
                int currSum = nums[i] + nums[left] + nums[right];
                int currDiff = abs(currSum - target);

                if (currDiff < bestDiff) {
                    bestDiff = currDiff;
                    bestSum = currSum;
                }

                if (currSum < target) {
                    left++;
                } else if (currSum > target) {
                    right--;
                } else {
                    return currSum;
                }
            }
        }

        return bestSum;
    }
};

int main() {
    Solution sol;

    std::vector<int> nums = {-1, 2, 1, -4};
    int target = 1;
    int result = sol.threeSumClosest(nums, target);
    
    std::cout << "Result: " << result << "\n";
    return 0;
}

File: 01.Python/0016.py
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n = len(nums)

        closestSum = 0
        closestDiff = float('inf')

        for i in range(n - 2):
            left = i + 1
            right = n - 1

            while left < right:
                currSum = nums[i] + nums[left] + nums[right]
                currDiff = abs(currSum - target)

                if currDiff < closestDiff:
                    closestDiff = currDiff
                    closestSum = currSum

                if currSum < target:
                    left += 1
                elif currSum > target:
                    right -= 1
                else:
                    return currSum

        return closestSum
    
sol = Solution()

nums = [-1,2,1,-4]
target = 1

result = sol.threeSumClosest(nums, target)
print(result)